2022 Performance chapter (#3092)
HTTPArchive/almanac.httparchive.org: src/content/en/2022/performance.md

By taking a hypothetical look at INP as a CWV, we can see just how much work there is to be done just to get back to FID-like levels. The most popular mobile websites would be especially affected by such a change as a consequence of their (over)use of JavaScript. Some CMSs and JavaScript frameworks would be hit harder than others, and it'll take an ecosystem-wide effort to collectively rein in the amount of client-side work that they do.

## Conclusion

As the industry continues to learn more about CWV, we're seeing steady improvement both in terms of implementation and across all top-level metric values themselves. The most visible performance optimization strides are at the platform level, like Android and bfcache improvements, given that their impact can be felt across many sites at once. But let's look at the most elusive piece of the performance puzzle: individual site owners.

> Google's decision to make CWV part of search ranking catapulted performance to the top of many companies' roadmaps, especially in the [SEO](./seo#core-web-vitals-cwv) industry. Individual site owners are certainly working hard to improve their performance and played a major role in the CWV improvements over the last year, even if those individual efforts are much harder to spot at this scale.

Reviewer: Is "played a major role" a bit strong given we can't back that up and have stated it was other changes outside of this that likely had the biggest impact?

WDYT about this:

```suggestion
Google's decision to make CWV part of search ranking catapulted performance to the top of many companies' roadmaps, especially in the [SEO](./seo#core-web-vitals-cwv) industry. Individual site owners are certainly working hard to improve their performance and likely played a major role in the CWV improvements over the last year, even if those individual efforts are much harder to spot at this scale.
```

Reviewer: Apparently discuss and agreed upon so we'll go with this!

authors: [mel-ada, rviscomi]
reviewers: [tunetheweb, pmeenan, 25prathamesh, estelle, konfirmed]
analysts: [rviscomi, 25prathamesh, siakaramalegos, konfirmed]
editors: [tunetheweb]
translators: []
results: https://docs.google.com/spreadsheets/d/1TPA_4xRTBB2fQZaBPZHVFvD0ikrR-4sNkfJfUEpjibs/
mel-ada_bio: Mel Ada is a software engineer on the Web Performance team at Etsy. Her current involvement in the community includes co-organizing the NY Web Performance Meetup and occaisonally doing some speaking of her own.
rviscomi_bio: Rick Viscomi is a DevRel engineer at Google, focusing on web performance. He is the co-author of <a hreflang="en" href="https://usingwpt.com/">Using WebPageTest</a>, a book about web performance testing. He also co-maintains HTTP Archive and is the Editor-in-Chief of the Web Almanac.
featured_quote: These results show that sites absolutely do have responsiveness issues, despite the rosy picture painted by FID. Regardless of whether INP becomes a CWV metric, your users will thank you if you start optimizing it now.
featured_stat_1: 39%

> featured_stat_label_1: Percent of LCP that is not statically discoverable

Reviewer: ```suggestion
featured_stat_label_1: LCP images that is not statically discoverable
```

featured_stat_2: 22%

> featured_stat_label_2: Percent of pages that are ineligble for bfache due to `no-store`

Reviewer: ```suggestion
featured_stat_label_2: Pages that are ineligble for bfache due to `no-store`
```

featured_stat_3: 20%

> featured_stat_label_3: Percent of the top 1k websites that would have good CWV with INP

Reviewer: ```suggestion
featured_stat_label_3: Top 1,000 websites that would have good CWV with INP
```

> unedited: true

Reviewer: ```suggestion
```

Enter <a hreflang="en" href="https://web.dev/vitals/">CWV</a>, three metrics to indicate overall performance health. Each metric covers an important area of user experience: loading, interactivity, and visual stability. The public <a hreflang="en" href="https://developer.chrome.com/docs/crux/">Chrome UX Report</a> (CrUX) dataset is Chrome's view of how millions of websites are performing on CWV. There's zero setup on the developer's part; Chrome automatically collects and publishes data from <a hreflang="en" href="https://developer.chrome.com/docs/crux/methodology/#eligibility">eligible</a> users and websites. Using CrUX data, we're able to get insights into the web's performance and even track historical data.

CWV offers a clearly defined approach to measuring how user's actually experience performance—a first for the industry. Are CWV the answer to helping the web become more performant? It's been a little over two years since launch, so this chapter examines where the web is currently with CWV and a look into the future.

> 

Reviewer: I think we should add a caveat about Rick's job at the env of the Introduction, similar to [how we've done in the other chapters](https://almanac.httparchive.org/en/2022/jamstack#:~:text=Disclosure%3A%20the%20two%20authors%20of%20this%20report%20were%20Netlify%20employees.%20Netlify%20invented%20the%20term%20Jamstack%20and%20owns%20Jamstack.org.%20This%20report%20and%20the%20underlying%20analysis%20were%20reviewed%20and%20approved%20by%20others%20not%20affiliated%20with%20Netlify.).

Author: Not sure if we want to name Rick or not here, the example didn't mention specific names but it sounds like all authors were included in the disclaimer. I'll leave that up to y'all!

Reviewer 2: Rephrasing slightly but LGTM

  description="Bar chart showing an image is the LCP content type on 82% of desktop pages and 72% of mobile pages, it is text on 17% and 26% respectively, and an inline image on 2% of desktop and 1% of mobile pages.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vR-dJP3uphZoGE5A_luniNBFm5V2ww6irfOxANg0hrMid7gjgrtchsN_utOIDOvVZUjIwpmUBb27nHF/pubchart?oid=839846485&format=interactive",
  sheets_gid="872701281",
  sql_file="lcp_resource_type.sql"
  )
}}

We see that 72% of mobile pages and 82% of desktop pages have images as their LCP. For example, these images may be traditional `img` elements or CSS background images. 26% of mobile pages and 17% of desktop pages have text-based content as their LCP.

1% of pages actually use inline images as their LCP content. This is almost always a bad idea for a number of reasons, mostly around caching and complexity.

#### LCP prioritization

After the HTML document is loaded, there are two major factors that affect how quickly the LCP resource itself can be loaded: discoverability and prioritization. We'll explore [LCP discoverability](#lcp-static-discoverability) later, but first let's look at how LCP images are prioritized.

> Images are not loaded at high priority by default, but thanks to the new <a hreflang="en" href="https://web.dev/priority-hints/">Priority Hints</a> API, developers can explicitly set their LCP images to load at high priority to take precedence over non-essential resources.

Reviewer: Images are not loaded _at all_ initially, never mind the their priority. This is an important point. The benefit of Priority Hints for LCP images isn't so much that they are boosted in priority, but that with that boost they are also eligible to be loaded in that early load phase.

> Images are not loaded at high priority by default, but thanks to the new <a hreflang="en" href="https://web.dev/priority-hints/">Priority Hints</a> API, developers can explicitly set their LCP images to load at high priority to take precedence over non-essential resources.

Reviewer 2: Is that just semantics, though? Like, what's the difference between saying that images are loaded after critical resources by default, and low priority resources are loaded after high priority resources? If images are low priority by default, setting them to high priority will make them load sooner with the other high priority resources.

For example, if there's a site with no CSS or JS and just an image, the image may be low priority by default, but it's not contending with anything else so it loads immediately, right?

> Images are not loaded at high priority by default, but thanks to the new <a hreflang="en" href="https://web.dev/priority-hints/">Priority Hints</a> API, developers can explicitly set their LCP images to load at high priority to take precedence over non-essential resources.

Reviewer: OK fair enough. I'm too in the weeds of this loading stuff :-)

Resolving.

{{ figure_markup(
  content="0.03%",
  caption="The percent of pages that use `fetchpriority=high` on their LCP element.",
  classes="big-number",
  sheets_gid="600760184",
  sql_file="lcp_element_data_2.sql"
  )
}}

0.03% of pages use `fetchpriority=high` on their LCP elements. Counterproductively, a handful of pages actually _lower_ the priority over their LCP images: 77 pages on mobile and 104 on desktop.

> `fetchpriority` is still very new and not supported everywhere, but there's little to no reason why every developer shouldn't be using it. [Patrick Meenan](https://twitter.com/patmeenan), who helped develop the API, [describes it](https://twitter.com/patmeenan/status/1460276602479251457) as a "cheat code" given how easy it is to implement relative to the potential improvements.

Reviewer: I disagree. See [this thread](https://webperformance.slack.com/archives/C04BK7K1X/p1666389089163239) where @philipwalton and I explain when it might not be a good idea to do it.

I think we should language like "there's little to no reason why every developer shouldn't be using it" when there clearer are cases of this.

> `fetchpriority` is still very new and not supported everywhere, but there's little to no reason why every developer shouldn't be using it. [Patrick Meenan](https://twitter.com/patmeenan), who helped develop the API, [describes it](https://twitter.com/patmeenan/status/1460276602479251457) as a "cheat code" given how easy it is to implement relative to the potential improvements.

Reviewer 2: WDYT about rephrasing to "there's little to no reason why it shouldn't be in every developer's toolbox"? The point I'm making is that everyone should use it responsibly, not necessarily that it should be blindly used on the same image, every time.

> `fetchpriority` is still very new and not supported everywhere, but there's little to no reason why every developer shouldn't be using it. [Patrick Meenan](https://twitter.com/patmeenan), who helped develop the API, [describes it](https://twitter.com/patmeenan/status/1460276602479251457) as a "cheat code" given how easy it is to implement relative to the potential improvements.

Reviewer: That works for me!

#### LCP static discoverability

Earlier we looked at [prioritization](#lcp-prioritization) as a way to improve LCP performance by outcompeting other concurrently loading resources. On the other hand, static discoverability helps to ensure that the LCP resource can be loaded as soon as possible.

An LCP image is considered to be _statically discoverable_ if its source URL can be parsed directly from the markup sent by the server. This definition includes sources defined within `picture` or `img` elements as well as sources that are explicitly preloaded.

One caveat is that text-based LCP content is always statically discoverable based on this definition. However, text-based content may sometimes depend on client-side rendering or web fonts, so consider these results as lower bounds.

> 

Reviewer 2: This feels out of place, prefer to keep with the next paragraph

> 

Reviewer: OK. I just thought CSR was the more likely reason, whereas it felt like an afterthought in current text. And the cusotm lazy-loading felts just dumped in there. But I'm fine keeping it as is if you prefer.

```html
<img data-src="kitten.jpg" class="lazyload">
```

> Custom lazy-loading techniques like the example above are one way that images are prevented from being statically discoverable, since they rely on JavaScript to update the `src` attribute. Client-side rendering may also obscure the LCP content.

Reviewer 2: Unnecessarily verbose IMO

{{ figure_markup(
  content="39%",
  caption="The percent of mobile pages on which the LCP element was not statically discoverable.",
  classes="big-number",
  sheets_gid="1465687616",
  sql_file="lcp_preload_discoverable.sql"
  )
}}

39% of mobile pages have LCP elements that are not statically discoverable. This figure is even worse on desktop at 44% of pages, potentially as a consequence of the previous section's finding that LCP content is more likely to be an image on desktop pages.

This is the first year that we're looking at this metric, so we don't have historical data for comparison, but these results hint at a big opportunity to improve the load delay of LCP resources.

#### LCP preloading

> When the LCP image is not statically discoverable, <a hreflang="en" href="https://web.dev/preload-critical-assets/">preloading</a> can be an effective way to minimize the load delay. Of course, it would be better if the resource was statically discoverable to begin with, but addressing that may require a complex rearchitecture of the way the page loads. Preloading is somewhat of a quick fix by comparison, as it can be implemented with a single HTTP header or `meta` tag.

Reviewer: Yeah fine. There's a couple of reasons why it's better but maybe a little too much to go into all those.

Note that the percentages in this section are out of only those pages in which the `img` element is the LCP, not all pages. For reference, recall that this accounts for 42% of pages.

{{ figure_markup(
  content="8.8%",
  caption="The percent of mobile pages that use custom lazy-loading on their LCP image.",
  classes="big-number",
  sheets_gid="1585533536",
  sql_file="lcp_lazy.sql"
  )
}}

One way that sites might polyfill the native lazy-loading behavior is to assign the image source to a `data-src` attribute and include an identifier like `lazyload` in the class list. Then, a script will watch the positions of images with this class name relative to the viewport, and swap the `data-src` value for the native `src` value to trigger the image to start loading.

Nearly as many pages are using this kind of custom lazy-loading behavior as native lazy-loading, at 8.8% of pages with `img`-based LCP.

> Beyond the adverse performance effects of lazy-loading LCP images, native image lazy-loaded is <a hreflang="en" href="https://caniuse.com/loading-lazy-attr">supported</a> by all major browsers, making custom solutions like these obsolete. In our opinion, developers should seriously consider removing these polyfills from their code and judiciously switching to use native lazy-loading.

Reviewer 2: Isn't obsolescence more of a fact, not an opinion, based on the browser support? The opinionated part is what developers should do about it.

> Beyond the adverse performance effects of lazy-loading LCP images, native image lazy-loaded is <a hreflang="en" href="https://caniuse.com/loading-lazy-attr">supported</a> by all major browsers, making custom solutions like these obsolete. In our opinion, developers should seriously consider removing these polyfills from their code and judiciously switching to use native lazy-loading.

Reviewer: As I mentioned previously some people still prefer custom solutions for the greater control they give - some feel Chrome is too aggressive in it's "near viewport" limits. So native lazy loading isn't an exact like-for-like replacement.

I agree with you it's my preferred way as don't believe the downsides (extra JS, inability to benefit from other Browser heurostics...etc.) outweigh this extra level of control. But it's still an opinion.

> Beyond the adverse performance effects of lazy-loading LCP images, native image lazy-loaded is <a hreflang="en" href="https://caniuse.com/loading-lazy-attr">supported</a> by all major browsers, making custom solutions like these obsolete. In our opinion, developers should seriously consider removing these polyfills from their code and judiciously switching to use native lazy-loading.

Reviewer 2: Ah got it, thanks for clarifying. Edited your suggestion to capture that nuance. WDYT?

> Beyond the adverse performance effects of lazy-loading LCP images, native image lazy-loaded is <a hreflang="en" href="https://caniuse.com/loading-lazy-attr">supported</a> by all major browsers, making custom solutions like these obsolete. In our opinion, developers should seriously consider removing these polyfills from their code and judiciously switching to use native lazy-loading.

Reviewer: LGTM!

        <td class="numeric">3%</td>
      </tr>
      <tr>
        <td>googleusercontent.com</td>
        <td class="numeric">2%</td>
        <td class="numeric">3%</td>
      </tr>
      <tr>
        <td>amazonaws.com</td>
        <td class="numeric">3%</td>
        <td class="numeric">3%</td>
      </tr>
      <tr>
        <td>wordpress.com</td>
        <td class="numeric">2%</td>
        <td class="numeric">2%</td>
      </tr>
      <tr>
        <td>cdn-website.com</td>
        <td></td>
        <td class="numeric">1%</td>
      </tr>
    </tbody>
  </table>
  <figcaption>
    {{ figure_link(
      caption="Top 10 hosts as a percent of pages with cross-origin LCP images.",
      sheets_gid="1432074792",
      sql_file="lcp_host_3p.sql"
    ) }}
  </figcaption>
</figure>

CMSs make up the most common hosts of cross-origin LCP images, which isn't too surprising given their popularity. What's interesting is that Shopify's domain accounts for 11% of all cross-origin images, even though it's only used by about 3% of websites, according to last year's [Ecommerce](../2021/ecommerce) chapter. Other CMSs like Wix, Blogger, WordPress, and Squarespace also appear in the list.

> It's promising that these CMSs are in control over where and how LCP images are served, but it could be a major effort to switch to a same-origin solution. Cross-origin preconnecting and preloading may be the next best things.

Reviewer 2: If we add this, we should have a source

> It's promising that these CMSs are in control over where and how LCP images are served, but it could be a major effort to switch to a same-origin solution. Cross-origin preconnecting and preloading may be the next best things.

Reviewer: Shopify and Wix use preconnects, and have started pushing them through Early Hints more recently: https://twitter.com/tunetheweb/status/1580889937242910720?s=20&t=5r2SmHls-06LedwFev5uWA or https://developer.chrome.com/blog/early-hints/

> It's promising that these CMSs are in control over where and how LCP images are served, but it could be a major effort to switch to a same-origin solution. Cross-origin preconnecting and preloading may be the next best things.

Reviewer 2: I'm starting to reconsider whether to include this host-specific analysis at all. You're right that if we're going to name specific hosts who are loading cross-origin LCP images, we should also give credit to the ones that are correctly mitigating those effects with resource hints. But that feels like too much of a tangent. WDYT?

> It's promising that these CMSs are in control over where and how LCP images are served, but it could be a major effort to switch to a same-origin solution. Cross-origin preconnecting and preloading may be the next best things.

Reviewer: I think you might be right. It's not great to do this, but as you say, the platforms do often mitigate it somewhat.

Bigger issue IMHO, is for smaller sites, that don't do this. But think that's covered enough with the section above (don't lose that btw!), but naming (and therefore shaming) the platforms here is probably not that helpful given that.

> It's promising that these CMSs are in control over where and how LCP images are served, but it could be a major effort to switch to a same-origin solution. Cross-origin preconnecting and preloading may be the next best things.

Reviewer 2: Ok, thanks, I'll remove this piece.

What we've seen so far is that render-blocking resources are still quite prevalent, very few sites are using advanced prioritization techniques, and more than a third of LCP images are not statically discoverable. Without concrete data to suggest that site owners or large publishing platforms are concertedly optimizing these aspects of their LCP performance, other places to look are optimizations at the OS or browser level.

According to a <a hreflang="en" href="https://blog.chromium.org/2022/03/a-new-speed-milestone-for-chrome.html#:~:text=Chrome%20continues%20to%20get%20faster%20on%20Android%20as%20well.%20Loading%20a%20page%20now%20takes%2015%25%20less%20time%2C%20thanks%20to%20prioritizing%20critical%20navigation%20moments%20on%20the%20browser%20user%20interface%20thread">Chromium blog post</a> in March 2022, loading performance on Android improved by 15%. The post doesn't go into too much detail, but it credits the improvement to "_prioritizing critical navigation moments on the browser user interface thread_." This may help explain why mobile performance outpaced desktop performance in 2022.

> The six percentage point improvement to LCP this year can only happen when hundreds of thousands of websites' performance improves. Putting aside the tantalizing question of how that happened, let's take a moment to appreciate that user experiences on the web _are_ getting better. It's hard work, but improvements like these make the ecosystem healthier and are worth celebrating.

Reviewer 2: Prefer to keep the emphasis

But how did bfcache improve CLS so much? Due to the way Chrome instantly restores the page from memory, its layout is settled and unaffected by any of the initial instability that typically occurs during loading.

One theory why LCP experiences didn't improve as dramatically is that back/forward navigations were already pretty fast thanks to standard HTTP caching. Remember, the threshold for "good" LCP is 2.5 seconds, which is pretty generous assuming any critical resources would already be in cache, and there are no bonus points for making "good" experiences even faster.

### CLS metadata and best practices

Let's explore how much of the web is adhering to CLS best practices.

#### Explicit dimensions

The most straightforward way to avoid layout shifts is to reserve space for content by setting explicit height and width attributes.

{{ figure_markup(
  content="72%",
  caption="The percent of mobile pages that fail to set explicit dimensions on at least one image.",
  classes="big-number",
  sheets_gid="1160188541",
  sql_file="cls_unsized_images.sql"
  )
}}

> 72% of mobile pages have unsized images. This stat alone shouldn't raise any alarms, because unsized images don't always result in user-perceived layout shifts, for example if they load outside of the viewport. Still, it's a sign that site owners may not be closely adhering to CLS best practices.

Reviewer 2: Keeping the "alarm" suggestion but I think the lazy-load scenario might be unnecessary.

#### bfcache eligibility

We've shown how impactful bfcache can be for CLS, so it's worth considering eligibility as a somewhat indirect best practice.

The best way to tell if a given page is eligible for bfcache is to <a hreflang="en" href="https://web.dev/bfcache/#test-to-ensure-your-pages-are-cacheable">test it in DevTools</a>. Unfortunately, there are over <a hreflang="en" href="https://docs.google.com/spreadsheets/d/1li0po_ETJAIybpaSX5rW_lUN62upQhY0tH4pR5UPt60/edit?usp=sharing">100 eligibility criteria</a>, many of which are hard or impossible to measure in the lab. So rather than looking at bfcache eligibility as a whole, let's look at a few criteria that are more easily measurable to get a sense for the lower bound of eligibility.

{{ figure_markup(
  image="bfcache-unload.png",
  caption="Usage of `unload` by site rank.",

>   description="Bar chart showing the percent of pages that are ineligble for bfcache due to setting unload handlers, grouped by site popularity rank. 36% of the top 1 thousand mobile pages set this handler, 33% of the top 10 thousand, 27% of the top 100 thousand, 21% of the top million, and 17% of all mobile pages. Desktop pages tend to use the unload handler slightly more often by a couple of percentage points across the ranks.",

Reviewer: ```suggestion
  description="Bar chart showing the percent of pages that are ineligble for bfcache due to setting `unload` handlers, grouped by site popularity rank. 36% of the top 1 thousand mobile pages set this handler, 33% of the top 10 thousand, 27% of the top 100 thousand, 21% of the top million, and 17% of all mobile pages. Desktop pages tend to use the `unload` handler slightly more often by a couple of percentage points across the ranks.",
```

22% of all mobile pages set this header, and 28% of mobile pages in the top 1k. This and the `unload` criteria are not mutually exclusive, so combined we can only say that at least 22% of mobile pages are ineligible for bfcache.

To reiterate, it's not that these things cause CLS issues. However, fixing them may make pages eligible for bfcache, which we've been shown to be an indirect yet powerful tool for improving layout stability.

### CLS conclusions

CLS is the CWV metric that improved the most in 2022 and it appears to have had a significant impact on the number of websites that have "good" overall CWV.

The cause of this improvement seems to come down to Chrome's launch of bfcache, which is reflected in the January 2022 CrUX dataset. However, at least a fifth of sites are ineligible for this feature due to aggressive `no-store` caching policies or discouraged use of the `unload` event listener. Correcting these anti-patterns is CLS's "one weird trick" to improve performance.

> There are other, more direct ways site owners can improve their CLS. Setting height and width attributes on images is the most straightforward one. Optimizing how animations are styled and how fonts load are two other—admittedly more complex—approaches to consider.

Reviewer: ```suggestion
There are other, more direct ways site owners can improve their CLS. Setting `height` and `width` attributes on images is the most straightforward one. Optimizing how animations are styled and how fonts load are two other—admittedly more complex—approaches to consider.
```

## First Input Delay (FID)

> <a hreflang="en" href="https://web.dev/fid/">First Input Delay</a> (FID) measures the time from the first user interaction like a click or tap to the time at which the browser begins processing the corresponding event handlers.
> 
> {{ figure_markup(
>   image="good-fid-by-device.png",
>   caption="Good FID by device",
>   description="Bar chart showing 100% of websites had good FID in 2020, 2021, and 2022. For sites visited on phones this increased from 80% in 2020 to 90% in 2021 and then to 92% in 2022.",
>   chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vR-dJP3uphZoGE5A_luniNBFm5V2ww6irfOxANg0hrMid7gjgrtchsN_utOIDOvVZUjIwpmUBb27nHF/pubchart?oid=1546220733&format=interactive",
>   sheets_gid="555510064",
>   sql_file="web_vitals_by_device.sql"
>   )
> }}
> 
> A website has "good" FID if at least 75 percent of all navigations across the site are faster than 100 ms. Effectively all websites have "good" FID for desktop users, and this trend has held firm over the years. Mobile FID performance is also exceptionally fast, with 92% of websites having "good" FID, a slight improvement over last year.

Reviewer: In previous sections you put the threshold before the graph.

```suggestion
<a hreflang="en" href="https://web.dev/fid/">First Input Delay</a> (FID) measures the time from the first user interaction like a click or tap to the time at which the browser begins processing the corresponding event handlers. A website has "good" FID if at least 75 percent of all navigations across the site are faster than 100 ms. 

{{ figure_markup(
  image="good-fid-by-device.png",
  caption="Good FID by device",
  description="Bar chart showing 100% of websites had good FID in 2020, 2021, and 2022. For sites visited on phones this increased from 80% in 2020 to 90% in 2021 and then to 92% in 2022.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vR-dJP3uphZoGE5A_luniNBFm5V2ww6irfOxANg0hrMid7gjgrtchsN_utOIDOvVZUjIwpmUBb27nHF/pubchart?oid=1546220733&format=interactive",
  sheets_gid="555510064",
  sql_file="web_vitals_by_device.sql"
  )
}}

Effectively all websites have "good" FID for desktop users, and this trend has held firm over the years. Mobile FID performance is also exceptionally fast, with 92% of websites having "good" FID, a slight improvement over last year.
```

> While it's great that so many websites have good FID experiences, developers need to be careful not to become too complacent. Google has been <a hreflang="en" href="https://web.dev/better-responsiveness-metric/">experimenting</a> with a new responsiveness metric that could end up replacing FID, which is especially important because sites tend to perform much worse on [this metric](./#interaction-to-next-paint-inp) than FID.

Reviewer: ```suggestion
While it's great that so many websites have good FID experiences, developers need to be careful not to become too complacent. Google has been <a hreflang="en" href="https://web.dev/better-responsiveness-metric/">experimenting with a new responsiveness metric</a> that could end up replacing FID, which is especially important because sites tend to perform much worse on [this new metric](./#interaction-to-next-paint-inp) than FID.
```

### Interaction to Next Paint (INP)

<a hreflang="en" href="https://web.dev/inp/">Interaction to Next Paint</a> (INP) measures the amount of time it takes for the browser to complete the next paint in response to a user interaction. This metric was created after Google <a hreflang="en" href="https://web.dev/responsiveness/">requested feedback</a> on a proposal to improve how we measure responsiveness. Many readers may be hearing about this metric for the first time, so it's worth going into a bit more detail about how it works.

An _interaction_ in this context refers to the user experience of providing an input to a web application and waiting for the next frame of visual feedback to be painted on the screen. The only inputs that are considered for INP are clicks, taps, and key presses. The INP value itself is taken from one of the worst interaction latencies on the page. Refer to the <a hreflang="en" href="https://web.dev/inp/#what-is-inp">INP documentation</a> for more info on how it's calculated.

> Unlike FID, INP is a measure of all interactions on the page, not just the first one. It also measures the entire time until the next frame is painted, unlike FID which only measures the time until the event handler only starts_ _processing. In these ways, INP is a much more representative metric of the holistic user experience on the page.

Reviewer: ```suggestion
Unlike FID, INP is a measure of all interactions on the page, not just the first one. It also measures the entire time until the next frame is painted, unlike FID which only measures the time until the event handler starts processing. In these ways, INP is a much more representative metric of the holistic user experience on the page.
```

A website has "good" INP if 75% of its INP experiences are faster than 200 ms. A website has "poor" INP if the 75th percentile is greater than or equal to 500 ms. Otherwise, it's INP is assessed as "needs improvement".

{{ figure_markup(
  image="inp-device.png",
  caption="Distribution of INP performance by device.",
  description="Bar chart showing the percent of websites having good, needs improvement, or poor INP by device. 95% of desktop websites have good INP, 4% need improvement, and 1% are poor. 55% of mobile websites have good INP, 36% need improvement, and 8% are poor.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vR-dJP3uphZoGE5A_luniNBFm5V2ww6irfOxANg0hrMid7gjgrtchsN_utOIDOvVZUjIwpmUBb27nHF/pubchart?oid=755106375&format=interactive",
  sheets_gid="555510064",
  sql_file="web_vitals_by_device.sql"
  )
}}

55% of websites have "good" INP on mobile, 36% are rated "needs improvement", and 8% have "poor" INP. The desktop story of INP is more similar to FID in that 95% of websites are rated "good", 4% are rated "needs improvement", and 1% are "poor".

> The enormous disparity between desktop and mobile users' INP experiences is much wider than with FID. This illustrates the extent to which mobile devices are struggling to keep up with the overwhelming amount of work websites do, and all signs point to the increasing reliance on <a hreflang="en" href="https://httparchive.org/reports/state-of-javascript?start=earliest&end=latest&view=grid#bytesJs">JavaScript</a> as a major factor.

Reviewer: Prefer Web Almanac over HTTP Archive resources?

```suggestion
The enormous disparity between desktop and mobile users' INP experiences is much wider than with FID. This illustrates the extent to which mobile devices are struggling to keep up with the overwhelming amount of work websites do, and all signs point to the increasing reliance on [JavaScript](./javascript#how-much-javascript-do-we-load) as a major factor.
```

Reviewer 2: I used the HA report in this case because it shows the change in JS over time

Reviewer: Fair enough - as long as there's a reason. Resolving.

{{ figure_markup(
  image="js-bytes-rank.png",
  caption="Median amount of JavaScript loaded per page, by rank.",
  description="Bar chart showing the median kilobytes of JavaScript per page, grouped by device and site rank. The median top 1 thousand mobile page loads 604 KB of JavaScript, 680 in the top 10 thousand, 610 in the top 100 thousand, 583 in the top million, and 462 overall. The values for desktop follow the same trend and are larger by tens of kilobytes.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vR-dJP3uphZoGE5A_luniNBFm5V2ww6irfOxANg0hrMid7gjgrtchsN_utOIDOvVZUjIwpmUBb27nHF/pubchart?oid=1725756250&format=interactive",
  sheets_gid="1423103831",
  sql_file="js_bytes_rank.sql"
  )
}}

When we look at the amount of JavaScript that the median mobile page loads for each of these ranks, it follows the same funny pattern! The median mobile page in the top 1k loads 604 KB of JavaScript, then it increases to 680 KB for the top 10k before dropping all the way down to 462 KB over all websites. These results don't prove that loading—and using—more JavaScript necessarily causes poor INP, but it definitely suggests a correlation exists.

> #### INP as a hypothetical CWV metric

Reviewer 3: I don't believe we're supposed to use CWV as an initialism. I'd change all instances of this to "Core Web Vitals".

Reviewer 2: Is that convention more for the official docs, or any post referencing CWV? I think given that we've used CWV in previous years' chapters and the community should easily recognize it, it's ok to keep, but not set on it.

Reviewer 2: Resolving for now to unblock publishing, but happy to take another look afterwards if needed

#### INP and TBT

What actually makes INP a better responsiveness metric than FID? One way to answer that question is to look at the correlation between field-based INP and FID performance and lab-based TBT performance. TBT is a direct measure of how unresponsive a page _can_ be, but it can never be a CWV itself because it doesn't actually measure the <a hreflang="en" href="https://web.dev/user-centric-performance-metrics/">user-perceived experience</a>.

This section draws from Annie Sullivan's <a hreflang="en" href="https://colab.sandbox.google.com/drive/12lJmAABgyVjaUbmWvrbzj9BkkTxw6ay2">research</a> using the May 2022 dataset.

{{ figure_markup(
  image="tbt-fid.png",
  alt="",

>   caption='Scatterplot visualizing the correlation between FID and TBT. (<a hreflang="en" href="https://colab.sandbox.google.com/drive/12lJmAABgyVjaUbmWvrbzj9BkkTxw6ay2">Source</a>))',

Reviewer: ```suggestion
  caption='Scatterplot visualizing the correlation between FID and TBT. (<a hreflang="en" href="https://colab.sandbox.google.com/drive/12lJmAABgyVjaUbmWvrbzj9BkkTxw6ay2">Source</a>)',
```

As Annie notes, both metrics are correlated with TBT, but she concludes that INP is more strongly correlated, making it a better responsiveness metric.

### FID conclusions

These results show that sites absolutely do have responsiveness issues, despite the rosy picture painted by FID. Regardless of whether INP becomes a CWV metric, your users will thank you if you start optimizing it now.

Nearly one in ten mobile sites are leaving free performance on the table by failing to disable double-tap to zoom. This is something all sites should be doing; it's only one line of HTML and it benefits both FID and INP. Run Lighthouse on your page and look for the <a hreflang="en" href="https://web.dev/viewport/">viewport</a> audit to be sure.

By taking a hypothetical look at INP as a CWV, we can see just how much work there is to be done just to get back to FID-like levels. The most popular mobile websites would be especially affected by such a change as a consequence of their (over)use of JavaScript. Some CMSs and JavaScript frameworks would be hit harder than others, and it'll take an ecosystem-wide effort to collectively rein in the amount of client-side work that they do.

## Conclusion

> As the industry continues to learn more about CWV, we're seeing steady improvement both in terms of implementation and the metric values themselves. The most visible performance optimization strides are at the browser and framework level where the impact can be felt across many sites at once. Teams like Aurora have set a precedent for collaborating with open source framework owners to introduce sweeping performance improvements. But let's look at the most elusive piece of the performance puzzle: individual site owners.

Reviewer: We don't really talk about Framework improvements in the chapter so think this paragraph should be re-written. If still including Aurora (I don't think we should given the contents of the chapter) then should link it.

```suggestion
As the industry continues to learn more about CWV, we're seeing steady improvement both in terms of implementation and the metric values themselves. The most visible performance optimization strides are at the browser and framework level where the impact can be felt across many sites at once. Teams like <a hreflang="en" href="https://developer.chrome.com/blog/introducing-aurora/">Aurora</a> have set a precedent for collaborating with open source framework owners to introduce sweeping performance improvements. But let's look at the most elusive piece of the performance puzzle: individual site owners.
```

Author: Passing on this suggestion as the content is no longer included.

> Google's decision to make CWV part of search ranking catapulted performance to the top of many companies' roadmaps, and we're seeing that reflected in the metrics.  Yet in some cases, developers choose to hack the metrics for SEO's sake rather than actually improve performance. There are as many reasons for this as there are stars in the sky, but one that we cannot ignore is how overwhelming and daunting web performance can be. We've also seen that there are some knowledge gaps to close with common causes of slow performance. The solution? Continue to invest in making web performance accessible. In the years to come, let's emphasize getting web performance knowledge the 'last mile' to the site owner.

Reviewer: Again rewrite this section as not covered now.
